Performance, Design and other fixes

emmanvg opened the pull request:
The major changes include:
- A new ``serialize()`` method to add flexibility for performance. You can pass the same arguments you would for a ``json.dumps()`` call. For example, `sort_keys=True`, `indent=4`, etc.
- Created a `CustomProperty` class, which developers can extend from to create their custom properties. It is wrapper, but it helps to signal the usage of custom properties. This fixes the problem when an object is parsed with `allow_custom=True` but the _properties dictionary would not get updated.
- ABC for `DataStore`, `DataSink` and `DataSource`. I think the changes made are good for proper concrete implementations of these ABCs. For example, DataStore was never abstract...

closes #99
closes #98
closes #96

clenk commented:
We talked about this offline, but just to document it: We also need to make sure custom properties' ordering works if they are passed to the constructor via a `custom_properties` dictionary instead of `allow_custom`. Like the 2nd example on https://stix2.readthedocs.io/en/latest/guide/custom.html#Custom-Properties.

mbastian1135 commented:
Why are we making the classes abstract now? (bc thats how I had them originally but then we decided not to have ABC's)

gtback commented:
@clenk: Is there a reason you think they would be handled differently? Everything gets collapsed into the same dictionary, whether they came from `custom_properties` dict or a standard kwarg (when `allow_custom=True`). It's interesting (and probably not desirable) that the custom properties come first in both examples on the Jupyter notebook you linked to.

gtback commented on stix2/base.py:
```diff
@@ -109,6 +102,13 @@ def __init__(self, allow_custom=False, **kwargs):
+            # The custom properties will get added to the bottom.
+            # Matched with a CustomProperty.
+            extra_kwargs = list(set(kwargs) - set(cls._properties))
+            self._properties.update([(x, CustomProperty()) for x in extra_kwargs])
```
I'm not sure we want to do this. It updates the class instance of `_properties`, meaning that any custom properties on the first instantiation of an SDO are allowed on any future instance, even if allow_custom is false for the future ones.

It's perhaps likely that people want to use the same custom properties on every instance of the same SDO that they create, but I'm not sure how I feel about the way this is done.

emmanvg replied:
That is a good point! So, could we (possible approaches):
- Make _properties an instance variable?
- Handle custom in a completely different structure?
- Go back with the old method (required checking the _inner dict)?

emmanvg replied:
I'll go back to the #85 fix. That was working fine.

gtback commented:
I'd like to figure out better ways to track the version history of the Jupyter notebooks. I noticed that a couple switched from Python 2 to Python 3 (which is OK, but I just want to make sure what we do is deliberate). @clenk , @mbastian1135 any ideas?

gtback commented on stix2/sources/__init__.py:
```diff
@@ -11,104 +11,106 @@
 
-    def get(self, stix_id, allow_custom=False):
+    @abstractmethod
+    def get(self, stix_id):  # pragma: no cover
```
Rather than making these abstract methods, can we just define them similar to `get(self, *args, **kwargs)`, and pass them on the source? I'd rather not require subclasses to override these.

clenk commented:
@gtback The notebooks will always switch Python version when someone re-executes cells in them unless we all use the exact same version of Python. I think we could use a git filter to ignore the lines about python version. See https://stackoverflow.com/q/16244969/ We could have git ignore the `execution_count` lines as well.

If you're referring to tracking changes in the content of output cells instead, there are some ways to handle that too: http://nbsphinx.readthedocs.io/en/0.2.16/usage.html#Using-Notebooks-with-Git.

gtback commented:
@clenk: I just want to make sure we don't accidentally introduce extraneous diffs in the Notebook JSON files (output, Python version, execution count, etc.). That link to the nbsphinx output is extremely useful, but do we really want to not pre-run any of the cells? Will that even work on ReadTheDocs? (sorry, more questions than answers, I know)

gtback commented on stix2/sources/__init__.py:
```diff
@@ -11,104 +11,106 @@
 
 
-class DataSink(object):
+class DataSink(with_metaclass(ABCMeta)):
```
Having DataSink and DataSource as abstract classes is OK. I just think DataStore can provide a lot of default functionality (delegating to `source` and `sink` attributes.

gtback commented on stix2/sources/filesystem.py:
```diff
@@ -35,6 +35,85 @@ def __init__(self, stix_dir, bundlify=False):
         self.source = FileSystemSource(stix_dir=stix_dir)
         self.sink = FileSystemSink(stix_dir=stix_dir, bundlify=bundlify)
 
+    def get(self, stix_id, allow_custom=False, version=None, _composite_filters=None):
```
None of this is specific to the FileSystem, I don't think. I'm hoping we can suitably abstract it with `*args` and `**kwargs` in the base class.

clenk commented:
As the Jupyter notebook versioning discussion seems tangential to this PR, I've moved it to #101.

emmanvg commented:
Hopefully I have addressed the concerns with the DataStores. It looks a lot cleaner now. Note: I did not change anything related to Jupyter Notebook since there is a new issue open. If nothing else needs to be fixed i'd say its ready to be merged! 👍
